File: laravel/app/Models/Content/Article.php
<?php

namespace App\Models\Content;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use App\Models\Players\Player;
use App\Models\Cards\Deck;

class Article extends Model
{
    protected $table = 'articles';

    protected $fillable = ['title', 'slug', 'body', 'excerpt', 'cover_image_url', 'status', 'article_type', 'language', 'view_count', 'likes_count', 'total_views_alltime', 'is_featured', 'published_at', 'author_id', 'featured_deck_id'];

    protected $casts = [
        'is_featured' => 'boolean',
        'published_at' => 'datetime',
    ];

    const STATUS_VALUES = ['Draft', 'Published', 'Archived'];
    const ARTICLE_TYPE_VALUES = ['Guide', 'Tierlist', 'Matchup', 'News', 'Spotlight', 'Decklist'];
    const LANGUAGE_VALUES = ['EN', 'DE', 'FR', 'IT', 'ES', 'JP', 'PT'];

    public function author(): BelongsTo
    {
        return $this->belongsTo(Player::class, 'author_id');
    }

    public function featuredDeck(): BelongsTo
    {
        return $this->belongsTo(Deck::class, 'featured_deck_id');
    }

    public function tagses(): BelongsToMany
    {
        return $this->belongsToMany(ArticleTag::class, 'article_tag_assignments', 'article_id', 'article_tag_id')->using(ArticleTagAssignment::class);
    }

    public function tagAssignments(): HasMany
    {
        return $this->hasMany(ArticleTagAssignment::class, 'article_id');
    }

    public function comments(): HasMany
    {
        return $this->hasMany(ArticleComment::class, 'article_id');
    }

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!(($this->view_count === null || $this->view_count >= 0))) {
            $errors['view_count_not_negative'] = 'Article view count must not be negative';
        }
        if (!(($this->likes_count === null || $this->likes_count >= 0))) {
            $errors['likes_count_not_negative'] = 'Article likes count must not be negative';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->status === 'Published' && $this->published_at === null) {
            throw new \RuntimeException('Published article must have a published_at timestamp');
        }
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    const ALLOWED_TRANSITIONS = [
        'Draft' => ['Published'],
        'Published' => ['Archived'],
        'Archived' => ['Draft'],
    ];

    public function assertTransition(string $to): void
    {
        $allowed = static::ALLOWED_TRANSITIONS[$this->status] ?? [];
        if (!in_array($to, $allowed, true)) {
            throw new \InvalidArgumentException("Transition {$this->status} -> {$to} not allowed");
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function publish(): void
    {
        // TODO: implement publish
    }

    public function archive(): void
    {
        // TODO: implement archive
    }

    public function replace($data): ?bool
    {
        // TODO: implement replace
        return null;
    }

    public function incrementView(): void
    {
        // TODO: implement increment_view
    }

    public function like(): void
    {
        // TODO: implement like
    }

    public function unlike(): void
    {
        // TODO: implement unlike
    }

    public function readingTimeMinutes(): ?int
    {
        // TODO: implement reading_time_minutes
        return null;
    }

    // ── Lifecycle hooks ──────────────────────────────────────────────
    protected static function boot(): void
    {
        parent::boot();
        static::saved(function (self $model) {
            $model->updateSearchIndex();
        });
    }

    protected function updateSearchIndex(): void
    {
        // TODO: implement update_search_index
    }

}

File: phoenix/lib/cards_project/content/article.ex
defmodule CardsProject.Content.Article do
  use Ecto.Schema
  import Ecto.Changeset

  schema "articles" do
    field :title, :string
    field :slug, :string
    field :body, :string
    field :excerpt, :string
    field :cover_image_url, :string
    field :status, :string
    field :article_type, :string
    field :language, :string
    field :view_count, :integer, default: 0
    field :likes_count, :integer, default: 0
    field :is_featured, :boolean, default: false
    field :published_at, :naive_datetime
    field :created_at, :naive_datetime
    belongs_to :author, CardsProject.Players.Player
    belongs_to :featured_deck, CardsProject.Cards.Deck
    belongs_to :comments, CardsProject.Content.ArticleComment
    many_to_many :tags, CardsProject.Content.ArticleTag, join_through: "article_tags_m2m"

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:title, :slug, :body, :view_count, :likes_count, :is_featured, :created_at, :excerpt, :cover_image_url, :status, :article_type, :language, :published_at, :author_id, :featured_deck_id])
    |> validate_required([:title, :slug, :body, :view_count, :likes_count, :is_featured, :created_at])
    |> validate_inclusion(:status, ["Draft", "Published", "Archived"])
    |> validate_inclusion(:article_type, ["Guide", "Tierlist", "Matchup", "News", "Spotlight", "Decklist"])
    |> validate_inclusion(:language, ["EN", "DE", "FR", "IT", "ES", "JP", "PT"])
    |> validate_number(:view_count, greater_than_or_equal_to: 0, message: "Article view count must not be negative")
    |> validate_number(:likes_count, greater_than_or_equal_to: 0, message: "Article likes count must not be negative")
    |> then(fn cs ->
      if get_field(cs, :status) == "Published" and (is_nil(get_field(cs, :published_at))) do
        Ecto.Changeset.add_error(cs, :published_at, "Published article must have a published_at timestamp")
      else
        cs
      end
    end)
  end

  # ── Business operations ────────────────────────────────────────────

  def publish(_record) do
    # TODO: implement Article.publish
    :ok
  end

  def archive(_record) do
    # TODO: implement Article.archive
    :ok
  end

  def increment_view(_record) do
    # TODO: implement Article.increment_view
    :ok
  end

  def like(_record) do
    # TODO: implement Article.like
    :ok
  end

  def unlike(_record) do
    # TODO: implement Article.unlike
    :ok
  end

  def reading_time_minutes(_record) do
    # TODO: implement Article.reading_time_minutes
    {:error, :not_implemented}
  end

  # ── Lifecycle state machine ─────────────────────────────────────────
  @allowed_transitions %{
    "Draft" => ["Published"],
    "Published" => ["Archived"],
    "Archived" => ["Draft"]
  }

  def assert_transition(%__MODULE__{status: current}, to) do
    allowed = Map.get(@allowed_transitions, current, [])
    if to in allowed do
      :ok
    else
      {:error, "Transition #{current} -> #{to} not allowed"}
    end
  end
end
